pkg/controllers/sharding/sharding_controller.go

```diff
@@ -0,0 +1,742 @@
+		workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName},
+	)
+	sc.nodeEventQueue = workqueue.NewTypedRateLimitingQueueWithConfig(
+		workqueue.DefaultTypedControllerRateLimiter[string](),
+		workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName + "-node-events"},
+	)
+
+	// Initialize components
+	// Initialize sharding manager
+	sc.assignmentCache = &AssignmentCache{
+		Assignments: make(map[string]*ShardAssignment),
+	}
+	sc.assignmentChangeChan = make(chan *AssignmentChangeEvent, 100)
+
+	// Parse scheduler configs from options
+	sc.parseSchedulerConfigsFromOptions()
+
+	// Initialize metrics cache
+	sc.nodeMetricsCache = make(map[string]*NodeMetrics)
+
+	// Initialize sharding manager with self as metrics provider
+	sc.shardingManager = NewShardingManager(sc.schedulerConfigs, sc)
+	sc.shardSyncPeriod = sc.controllerOptions.ShardSyncPeriod
+
+	// Setup event handlers
+	sc.nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sc.addNodeEvent,
+		UpdateFunc: sc.updateNodeEvent,
+		DeleteFunc: sc.deleteNodeEvent,
+	})
+
+	sc.shardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sc.addShard,
+		UpdateFunc: sc.updateShard,
+		DeleteFunc: sc.deleteShard,
+	})
+
+	sc.podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sc.addPod,
+		UpdateFunc: sc.updatePod,
+		DeleteFunc: sc.deletePod,
+	})
+	klog.V(6).Infof("ShardingController initialized successfully")
+
+	return nil
+}
+
+// Run starts the controller
+func (sc *ShardingController) Run(stopCh <-chan struct{}) {
+	defer sc.nodeShardQueue.ShutDown()
+	defer sc.nodeEventQueue.ShutDown()
+
+	klog.Infof("Starting sharding controller.")
+	defer klog.Infof("Shutting down sharding controller.")
+
+	// Start informer factories
+	sc.kubeInformerFactory.Start(stopCh)
+	sc.vcInformerFactory.Start(stopCh)
+
+	// Add specific sync checks with detailed logging
+	klog.Infof("Waiting for cache synchronization...")
+	if !cache.WaitForCacheSync(
+		stopCh,
+		sc.nodeInformer.Informer().HasSynced,
+		sc.shardInformer.Informer().HasSynced,
+		sc.podInformer.Informer().HasSynced,
+	) {
+		klog.Errorf("cache sync failed")
+	}
+	klog.Infof("Cache synchronization completed successfully")
+
+	// Initialize node metrics
+	sc.refreshNodeMetrics()
+
+	// Initial sync
+	sc.syncShards()
+
+	// Start workers
+	go wait.Until(sc.worker, time.Second, stopCh)
+	go wait.Until(sc.nodeEventWorker, time.Second, stopCh)
+
+	// Start periodic sync
+	go wait.Until(sc.syncShards, sc.shardSyncPeriod, stopCh)
+
+	// Start assignment change processor
+	go sc.assignmentChangeProcessor(stopCh)
+
+	// Start periodic metrics refresh
+	go wait.Until(sc.refreshNodeMetrics, nodeRefreshPeriod, stopCh)
+
+	<-stopCh
+	klog.Infof("Shutting down %s", controllerName)
+}
+
+// refreshNodeMetrics periodically refreshes node metrics
+func (sc *ShardingController) refreshNodeMetrics() {
```

Comment:
Is it necessary to refresh all the node metrics periodically? I saw that sharding controller will refresh node metrics according to the node events